GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,195 @@
+            }
+
+            return this.queue.TryAdd(message);
+        }
+
+        public void Stop()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (this.queue.IsAddingCompleted)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            this.queue.CompleteAdding();
+            this.writerThread.Join();
+
+            Debug.Assert(this.queue.IsCompleted, "Message queue should be empty after being stopped");
+        }
+
+        public void Dispose()
+        {
+            if (!this.isDisposed)
+            {
+                this.Stop();
+
+                this.pipeClient?.Dispose();
+                this.pipeClient = null;
+
+                this.writerThread = null;
+
+                this.queue.Dispose();
+            }
+
+            this.isDisposed = true;
+        }
+
+        private void RaiseStateChanged(QueuedPipeStringWriterState newState, Exception ex)
+        {
+            if (this.state != newState)
+            {
+                this.state = newState;
+                this.eventSink?.OnStateChanged(this, newState, ex);
+            }
+        }
+
+        private void ThreadProc()
+        {
+            // Drain the queue of all messages currently in the queue.
+            // TryTake() using an infinite timeout will block until either a message is available (returns true)
+            // or the queue has been marked as completed _and_ is empty (returns false).
+            string message;
+            while (this.queue.TryTake(out message, Timeout.Infinite))
+            {
+                if (message != null)
+                {
+                    this.WriteMessage(message);
+                }
+            }
+
+            this.RaiseStateChanged(QueuedPipeStringWriterState.Stopped, null);
+        }
+
+        private void WriteMessage(string message)
+        {
+            // Create pipe if this is the first message, or if the last connection broke for any reason
+            if (this.pipeClient == null)
+            {
+                try
+                {
+                    // Create a new pipe stream instance using the provided factory
+                    NamedPipeClientStream pipe = this.createPipeFunc();
+
+                    // Specify a instantaneous timeout because we don't want to hold up the
+                    // background thread loop if the pipe is not available; we will just drop this event.
+                    // The pipe server should already be running and waiting for connections from us.
+                    pipe.Connect(timeout: 0);
+
+                    // Keep a hold of this connected pipe for future messages
+                    this.pipeClient = pipe;
+                }
+                catch (Exception ex)
```

Comment:
Is there a more specific set of expected exceptions we could catch here?

Reply:
The mantra of the telemetry-over-pipe tracing was to ensure we _never ever_ brought down the mount process just because of a problem with telemetry. Yes, we expect only 'TimeoutException' or maybe 'BrokenPipe' (or maybe some permission/security exceptions too).

The comment raised by @wilbaker about the `ThreadProc` method and calling `Environment.Exit(1)` on `catch` to bring down the entire process is an open question here, waiting on @jrbriggs's input about if the "don't crash at all costs" idea is still valid 😃